x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java

```diff
@@ -141,13 +143,22 @@ protected void readInternal(ByteBuffer b) throws IOException {
             try {
                 final CacheFile cacheFile = getCacheFileSafe();
                 try (Releasable ignored = cacheFile.fileLock()) {
-                    final Tuple<Long, Long> range = computeRange(pos);
-                    bytesRead = cacheFile.fetchRange(
-                        range.v1(),
-                        range.v2(),
-                        (start, end) -> readCacheFile(cacheFile.getChannel(), end, pos, b, len),
-                        (start, end) -> writeCacheFile(cacheFile.getChannel(), start, end)
-                    ).get();
+                    final Tuple<Long, Long> rangeToWrite = computeRange(pos);
+                    final Tuple<Long, Long> rangeToRead = Tuple.tuple(pos, Math.min(pos + len, rangeToWrite.v2()));
+
+                    bytesRead = cacheFile.fetchAsync(rangeToWrite, rangeToRead, (channel) -> {
+                        final int read;
+                        if ((rangeToRead.v2() - rangeToRead.v1()) < b.remaining()) {
```

Comment:
This was previously embedded in the `readCacheFile()` method but feels wrong to me; the `readCacheFile()` should expect a `ByteBuffer` with an appropriate limit.

Reply:
++ there was a recent change to the Lucene APIs replacing `byte[], int, int` with `ByteBuffer` but this was not pushed all the way through this code.